examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,609 @@
+affects the quality of the produced encoders, but more examples would lead to
+longer training time.
+"""
+
+train_size = 30000
+captions_per_image = 2
+images_per_file = 2000
+train_image_paths = image_paths[:train_size]
+num_train_files = int(np.ceil(train_size / images_per_file))
+train_files_prefix = os.path.join(tfrecords_dir, "train")
+
+if tf.io.gfile.exists(tfrecords_dir):
+    print("Removing previous tfrecord files...")
+    tf.io.gfile.rmtree(tfrecords_dir)
+tf.io.gfile.makedirs(tfrecords_dir)
+print("New tfrecord directory created.")
+
+
+def bytes_feature(value):
+    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))
+
+
+def create_example(image_path, caption):
+    raw_image = tf.io.read_file(image_path).numpy()
+    feature = {
+        "caption": bytes_feature(caption.encode()),
+        "raw_image": bytes_feature(raw_image),
+    }
+    example = tf.train.Example(features=tf.train.Features(feature=feature))
+    return example
+
+
+def write_tfrecords(file_name, image_paths):
+    print(f"Writing {file_name}...")
+    caption_list = []
+    image_path_list = []
+    for image_path in image_paths:
+        captions = image_path_to_caption[image_path][:captions_per_image]
+        caption_list.extend(captions)
+        image_path_list.extend([image_path] * len(captions))
+
+    with tf.io.TFRecordWriter(file_name) as writer:
+        for example_idx in range(len(image_path_list)):
+            example = create_example(
+                image_path_list[example_idx], caption_list[example_idx]
+            )
+            writer.write(example.SerializeToString())
+    return example_idx + 1
+
+
+def write_data(image_paths, num_files, files_prefix):
+    example_counter = 0
+    for file_idx in range(num_files):
+        file_name = files_prefix + "-%02d.tfrecord" % (file_idx)
+        start_idx = images_per_file * file_idx
+        end_idx = start_idx + images_per_file
+        current_image_paths = image_paths[start_idx:end_idx]
+        example_counter += write_tfrecords(file_name, current_image_paths)
+        print(f"{example_counter} examples were written.")
+    return example_counter
+
+
+print("Writing training tfrecords...")
+train_example_count = write_data(train_image_paths, num_train_files, train_files_prefix)
+print("Training tfrecord files created successfully.")
+
+"""
+### Create `tf.data.Dataset` for training and evaluation
+"""
+
+AUTOTUNE = tf.data.experimental.AUTOTUNE
+
+feature_description = {
+    "caption": tf.io.FixedLenFeature([], tf.string),
+    "raw_image": tf.io.FixedLenFeature([], tf.string),
+}
+
+
+def read_example(example):
+    features = tf.io.parse_single_example(example, feature_description)
+    raw_image = features.pop("raw_image")
+    image = tf.image.decode_jpeg(raw_image, channels=3)
+    image = tf.image.resize(image, (299, 299))
+    features["image"] = image
+    return features, tf.constant(1)
```

Comment:
The `model.fit` expects a tuple of (inputs, targets). Since the targets are dynamically computed during loss calculation, a tf constant is added as a placeholder for the targets. I added this a comment for clarification.